File: 2024/08.py
#!/usr/bin/python3.12

from pyaoc import Env, Grid
from collections import defaultdict

e = Env(8)
e.T("""............
........0...
.....0......
.......0....
....0.......
......A.....
............
............
........A...
.........A..
............
............""", 14, 34)

e.T("""..........
..........
..........
....a.....
........a.
.....a....
..........
......A...
..........
..........""", 4, None)

e.T("""T.........
...T......
.T........
..........
..........
..........
..........
..........
..........
..........""", 3, 9)


def find_antennas(g):
    antennas = defaultdict(set)
    for row in range(g.h):
        for col in range(g.w):
            c = g.get(row, col)
            if c != '.':
                antennas[c].add((row, col))
    return antennas


def antinodes_of_kind(g, positions, antinodes):
    for i in range(len(positions)):
        first = positions[i]
        for j in range(i+1, len(positions)):
            # For each pair of positions
            second = positions[j]
            dr = second[0] - first[0]
            dc = second[1] - first[1]
            right = (second[0] + dr, second[1] + dc)
            left = (first[0] - dr, first[1] - dc)
            if g.is_in(right[0], right[1]):
                antinodes.add(right)
            if g.is_in(left[0], left[1]):
                antinodes.add(left)


def part1(input):
    g = Grid(input.get_valid_lines())
    ant = find_antennas(g)
    antinodes = set()
    for _, s in ant.items():
        antinodes_of_kind(g, list(s), antinodes)
    return len(antinodes)


e.run_tests(1, part1)
e.run_main(1, part1)


def antinodes_with_resonance(g, positions, antinodes):
    for i in range(len(positions)):
        first = positions[i]
        for j in range(i+1, len(positions)):
            # For each pair of positions
            second = positions[j]
            dr = second[0] - first[0]
            dc = second[1] - first[1]
            pos = first
            while g.is_in(pos[0], pos[1]):
                antinodes.add(pos)
                pos = (pos[0] + dr, pos[1] + dc)
            pos = (first[0] - dr, first[1] - dc)
            while g.is_in(pos[0], pos[1]):
                antinodes.add(pos)
                pos = (pos[0] - dr, pos[1] - dc)


def part2(input):
    g = Grid(input.get_valid_lines())
    ant = find_antennas(g)
    antinodes = set()
    for _, s in ant.items():
        antinodes_with_resonance(g, list(s), antinodes)
    return len(antinodes)


e.run_tests(2, part2)
e.run_main(2, part2)

File: 2024/ts/08.ts
type Pos = [number, number];

class Grid {
    width: number = 0;
    height: number = 0;
    antennas: Map<string, Pos[]> = new Map();
}

const parse_input = (raw_input: string): Grid => {
    const grid = new Grid();
    let row = 0;
    for (let ln of raw_input.split('\n')) {
        ln = ln.trim();
        if (!ln) {
            break;
        }
        if (grid.width === 0) {
            grid.width = ln.length;
        } else {
            console.assert(grid.width === ln.length, `rows of different lengths: ${grid.width} and ${ln.length}`);
        }
        grid.height++;
        for (let i = 0; i < grid.width; ++i) {
            const c = ln[i];
            if (c !== '.') {
                if (!grid.antennas.has(c)) {
                    grid.antennas.set(c, []);
                }
                grid.antennas.get(c)?.push([row, i]);
            }
        }
        row++;
    }
    return grid;
}

const get_antenna_pairs = (grid: Grid): [Pos, Pos][] => {
    const pairs: [Pos, Pos][] = [];
    for (const points of grid.antennas.values()) {
        for (let i = 0; i < points.length; ++i) {
            for (let j = i + 1; j < points.length; ++j) {
                pairs.push([points[i], points[j]]);
            }
        }
    }
    return pairs;
}

const find_antinodes = (grid: Grid, resonance: boolean): Set<number> => {
    const pairs = get_antenna_pairs(grid);
    // Note - using a set of numbers, where each number is row*width+col.
    // Cannot use tuples/pairs of [row, col], because Set keys are compared
    // by object reference equality, not by values.
    const antinodes = new Set<number>();
    for (const [first, second] of pairs) {
        const d_row = second[0] - first[0];
        const d_col = second[1] - first[1];
        let nr = first[0] - d_row;
        let nc = first[1] - d_col;
        while (nr >= 0 && nc >= 0 && nr < grid.height && nc < grid.width) {
            antinodes.add(nr * grid.width + nc);
            if (!resonance) {
                break;
            }
            nr -= d_row;
            nc -= d_col;
        }
        nr = second[0] + d_row;
        nc = second[1] + d_col;
        while (nr >= 0 && nc >= 0 && nr < grid.height && nc < grid.width) {
            antinodes.add(nr * grid.width + nc);
            if (!resonance) {
                break;
            }
            nr += d_row;
            nc += d_col;
        }
        if (resonance) {
            antinodes.add(first[0] * grid.width + first[1]);
            antinodes.add(second[0] * grid.width + second[1]);
        }
    }
    return antinodes;
}

const part1 = (raw_input: string): number => {
    const grid = parse_input(raw_input);
    const antinodes = find_antinodes(grid, false);
    return antinodes.size;
}

const part2 = (raw_input: string): number => {
    const grid = parse_input(raw_input);
    const antinodes = find_antinodes(grid, true);
    return antinodes.size;
}

export {
    Grid,
    find_antinodes,
    get_antenna_pairs,
    parse_input,
    part1,
    part2,
}
